### PR TITLE
hex: Add from_hex<T> for user types

### PR DESCRIPTION
**Pull request by chfast**
Depends on #648.

### CODE REVIEW

File: include/evmc/hex.hpp [hunk 1]
```diff
@@ -113,4 +113,24 @@ inline std::optional<bytes> from_hex(std::string_view hex)
         return {};
     return bs;
 }
+
+/// Decodes hex-encoded string into custom type T with .bytes array of uint8_t.
+///
+/// The decoded bytes are right aligned in case the input is smaller than the result type.
```
Comment (axic) [inline include/evmc/hex.hpp:119]: Why?

Comment (chfast) [reply include/evmc/hex.hpp:119]: Because they imitate BE numbers.

Comment (chfast) [reply include/evmc/hex.hpp:119]: Yes, but left/right does not have clear meaning.

Comment (chfast) [reply include/evmc/hex.hpp:119]: > When the input is smaller than the result type, the result is padded with zeros on the left (the result bytes of lowest indices are filled with zeros).

?

Comment (chfast) [reply include/evmc/hex.hpp:119]: Yes, an option can be added if needed, but so far this is only used for literals and loading storage values from JSON.

**Review by gumb0 — APPROVED** [review]

File: include/evmc/hex.hpp (hunk 1 above)
Comment (axic) [reply include/evmc/hex.hpp:119]: Perhaps add `// TODO: support left alignment too with a flag`.

**Review by axic — APPROVED** [review]

**Review by axic — COMMENTED** [review]
Perhaps add `// TODO: support left alignment too with a flag`.

**Pull request merged**